scala:console crashes with scala 2.12.0

Scala console crashes shortly after starting up for a project depending on scala 2.12.0:
```
/t/tmpsh.64445 ▶ mvn scala:console
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building test NEXT-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- scala-maven-plugin:3.2.2:console (default-cli) @ test ---
[WARNING] scala-maven-plugin cannot fork scala console!!  Running in process
Welcome to Scala 2.12.0 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_112).
Type in expressions for evaluation. Or try :help.

scala> [INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 6.101 s
[INFO] Finished at: 2016-11-16T18:46:39+03:00
[INFO] Final Memory: 49M/597M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:3.2.2:console (default-cli) on project test: wrap: java.lang.reflect.InvocationTargetException: jline.console.completer.CandidateListCompletionHandler.setPrintSpaceAfterFullCompletion(Z)V -> [Help 1]
…
```
Here's what POM looks like:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>test</groupId>
    <artifactId>test</artifactId>
    <version>NEXT-SNAPSHOT</version>

    <dependencies>
        <dependency>
            <groupId>org.scala-lang</groupId>
            <artifactId>scala-library</artifactId>
            <version>2.12.0</version>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>net.alchim31.maven</groupId>
                <artifactId>scala-maven-plugin</artifactId>
                <version>3.2.2</version>
                <executions>
                    <execution>
                        <goals>
                            <goal>compile</goal>
                            <goal>testCompile</goal>
                            <goal>add-source</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```
When scala version is set to 2.11.8, the consle works fine. I've also attached the [output](https://github.com/davidB/scala-maven-plugin/files/596892/maven-output.txt)
 of `mvn -X scala:console`. I'm not quite sure this is the plugin's issue, but the standalone scala console works fine for me.

--- a/src/main/java/scala_maven/ScalaConsoleMojo.java
+++ b/src/main/java/scala_maven/ScalaConsoleMojo.java
@@ -1,7 +1,9 @@
 package scala_maven;
 
 import java.io.File;
+import java.io.IOException;
 import java.util.ArrayList;
+import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Set;
@@ -9,6 +11,8 @@ import java.util.Set;
 import scala_maven_executions.JavaMainCaller;
 import scala_maven_executions.JavaMainCallerInProcess;
 import scala_maven_executions.MainHelper;
+
+import org.apache.maven.artifact.Artifact;
 
 /**
  * Run the Scala console with all the classes of the projects (dependencies and builded)
@@ -20,6 +24,22 @@ import scala_maven_executions.MainHelper;
  * @executionStrategy once-per-session
  */
 public class ScalaConsoleMojo extends ScalaMojoSupport {
+
+    // Private Static Values //
+
+    /**
+     * Constant {@link String} for "jline". Used for the artifact id, and
+     * usually the group id, for the JLine library needed by the Scala Console.
+     */
+    private static final String JLINE = "jline";
+
+    /**
+     * Constant {@link String} for "org.scala-lang". In this class it is used
+     * for the forked JLine group id.
+     */
+    private static final String SCALA_ORG_GROUP = "org.scala-lang";
+
+    // Instance Members //
 
     /**
      * The console to run.
@@ -55,50 +75,243 @@ public class ScalaConsoleMojo extends ScalaMojoSupport {
 
     @Override
     protected void doExecute() throws Exception {
-        //TODO - Many other paths uses the getScalaCommand()!!! We should try to use that as much as possibel to help maintainability.
-        VersionNumber scalaVersion = findScalaVersion();
-        Set<String> classpath = new LinkedHashSet<String>();
-        addCompilerToClasspath(classpath);
-        addLibraryToClasspath(classpath);
-        if (new VersionNumber("2.11.0").compareTo(scalaVersion) <= 0) {
-            addToClasspath("jline", "jline", "2.12", classpath);
-        } else if (new VersionNumber("2.9.0").compareTo(scalaVersion) <= 0) {
-          addToClasspath("org.scala-lang", "jline", scalaVersion.toString(), classpath);
-        } else {
-          addToClasspath("jline", "jline", "0.9.94", classpath);
-        }
-        classpath.addAll(project.getCompileClasspathElements());
-        if (useTestClasspath) {
-            classpath.addAll(project.getTestClasspathElements());
-        }
-        if (useRuntimeClasspath) {
-            classpath.addAll(project.getRuntimeClasspathElements());
-        }
-        String classpathStr = MainHelper.toMultiPath(classpath.toArray(new String[classpath.size()]));
-        JavaMainCaller jcmd = null;
-        List<String> list = new ArrayList<String>(args != null ? args.length + 3 : 3);
-        if(args != null) {
-            for(String arg : args) {
-                list.add(arg);
-            }
-        }
-        list.add("-cp");
-        list.add(classpathStr);
+        // Force no forking
+        final JavaMainCaller jcmd = super.getScalaCommand(false, this.mainConsole);
+        // Determine Scala Version
+        final VersionNumber scalaVersion = super.findScalaVersion();
+        final Set<String> classpath = this.setupClassPathForConsole(scalaVersion);
 
-        if(fork) {
-            getLog().warn("scala-maven-plugin cannot fork scala console!!  Running in process");
+        // Log if we are violating the user settings.
+        if (super.fork) {
+            super.getLog().info("Ignoring fork for console execution.");
         }
 
-        jcmd = new JavaMainCallerInProcess(this, mainConsole, classpathStr, jvmArgs, list.toArray(new String[list.size()]));
-        //We need to make sure compiler plugins are sent into the interpreter as well!
-        addCompilerPluginOptions(jcmd);
-        if (javaRebelPath != null) {
-            if (!javaRebelPath.exists()) {
-                getLog().warn("javaRevelPath '"+javaRebelPath.getCanonicalPath()+"' not found");
+        // Setup the classpath
+
+
+        // Build the classpath string.
+        final String classpathStr = MainHelper.toMultiPath(classpath.toArray(new String[classpath.size()]));
+
+        // Setup the JavaMainCaller
+        jcmd.addArgs(super.args);
+        jcmd.addOption("-cp", classpathStr);
+        super.addCompilerPluginOptions(jcmd);
+
+        // Check for Java Rebel
+        this.handleJavaRebel(jcmd);
+
+        // Run
+        jcmd.run(super.displayCmd);
+    }
+
+    // Private Methods //
+
+    /**
+     * If {@link #javaRebelPath} is defined, then attempt to resolve it on the
+     * filesystem and setup the Scala console to use it.
+     * <p>
+     * If we are unable to find it or {@link #javaRebelPath} is {@code null},
+     * then nothing is done to the given {@link JavaMainCaller}.
+     *
+     * @param jcmd the {@link JavaMainCaller} to which to add the JRebel settings.
+     */
+    private void handleJavaRebel(final JavaMainCaller jcmd) throws IOException {
+        if (this.javaRebelPath != null) {
+            final String canonicalJavaRebelPath = this.javaRebelPath.getCanonicalPath();
+            if (this.javaRebelPath.exists()) {
+                jcmd.addJvmArgs("-noverify", String.format("-javaagent:%s", canonicalJavaRebelPath));
             } else {
-                jcmd.addJvmArgs("-noverify", "-javaagent:" + javaRebelPath.getCanonicalPath());
+                super.getLog().warn(String.format("javaRevelPath '%s' not found", canonicalJavaRebelPath));
             }
         }
-        jcmd.run(displayCmd);
+    }
+
+    /**
+     * Construct the appropriate Classpath for the Scala console.
+     *
+     * @param scalaVersion the {@link VersionNumber} for the Scala
+     *                     Compiler/Library we are using.
+     *
+     * @return A {@link Set} of {@link String} defining classpath values to
+     *         provide to the Scala console.
+     *
+     * @throws {@link Exception} for many reasons, mostly relating to ad-hoc
+     *         dependency resolution.
+     */
+    private Set<String> setupClassPathForConsole(final VersionNumber scalaVersion) throws Exception {
+        final Set<String> classpath = new HashSet<String>();
+
+        classpath.addAll(this.setupProjectClasspaths());
+        classpath.addAll(this.setupConsoleClasspaths(scalaVersion));
+
+        return classpath;
+    }
+
+    /**
+     * Construct the Classpath defined by the project and plugin settings.
+     * <p>
+     * This should include the following entities.
+     * <ul>
+     *   <li> Scala Compiler, as defined by {@link ScalaMojoSupport#addCompilerToClasspath}.
+     *   <li> Library Classpath, as defined by {@link ScalaMojoSupport#addLibraryToClasspath}.
+     *   <li> Test Classpath, if {@link #useTestClasspath} is {@code true}.
+     *   <li> Runtime Classpath, if {@link #useRuntimeClasspath} is {@code true}.
+     * </ul>
+     *
+     * @return A {@link Set} of {@link String} defining the classpath for the
+     *         project and plugin settings.
+     *
+     * @throws {@link Exception} for many reasons, mostly relating to ad-hoc
+     *         dependency resolution.
+     */
+    private Set<String> setupProjectClasspaths() throws Exception {
+        final Set<String> classpath = new HashSet<String>();
+
+        super.addCompilerToClasspath(classpath);
+        super.addLibraryToClasspath(classpath);
+
+        if (this.useTestClasspath) {
+            classpath.addAll(super.project.getTestClasspathElements());
+        }
+
+        if (this.useRuntimeClasspath) {
+            classpath.addAll(super.project.getRuntimeClasspathElements());
+        }
+
+        return classpath;
+    }
+
+    /**
+     * Construct the Classpath for any additional dependencies that are needed
+     * to run the Scala console.
+     * <p>
+     * This should include the following entities.
+     * <ul>
+     *   <li> Jline, used for readline like features in the REPL.
+     * </ul>
+     *
+     * @param scalaVersion the version of the Scala Compiler/Library we are
+     *        using for this execution.
+     *
+     * @return A {@link Set} of {@link String} of the classpath as defined by
+     *
+     * @throws {@link Exception} for many reasons, mostly relating to ad-hoc
+     *         dependency resolution.
+     */
+    private Set<String> setupConsoleClasspaths(final VersionNumber scalaVersion) throws Exception {
+        final Set<String> classpath = new HashSet<String>();
+
+        super.addToClasspath(this.resolveJLine(scalaVersion, this.fallbackJLine(scalaVersion)), classpath, true);
+
+        return classpath;
+    }
+
+    /**
+     * Attempt to resolve JLine against the Scala Compiler's dependency tree.
+     * <p>
+     * This allows us to not have to worry about manually keeping the JLine
+     * dependency synchronized with the Scala REPL, which is likely to cause
+     * binary compatibility errors. If, for some reason, we are unable to
+     * resolve this dependency using the Compiler's dependency tree, we fallback
+     * to a set of hard-coded defaults that will <em>usually</em> work, but will
+     * not always work.
+     * <p>
+     * If the dynamic approach to finding the JLine dependency proves stable, we
+     * may drop the fallback in the future.
+     *
+     * @param scalaVersion the version of the Scala Compiler/Library we are
+     *        using for this execution.
+     * @param defaultFallback returned if we are unable to resolve JLine against
+     *        the Scala Compiler's dependency tree.
+     *
+     * @return an {@link Artifact} to provide to the runtime of the Scala
+     *         console conforming the JLine.
+     *
+     * @throws {@link Exception} for many reasons, mostly relating to ad-hoc
+     *         dependency resolution.
+     */
+    private Artifact resolveJLine(final VersionNumber scalaVersion,
+                                  final Artifact defaultFallback) throws Exception {
+        final Artifact compilerArtifact =
+            super.scalaCompilerArtifact(scalaVersion.toString());
+        final Set<Artifact> compilerDeps =
+            super.resolveArtifactDependencies(compilerArtifact);
+        for (final Artifact a : compilerDeps) {
+            super.getLog().warn(a.toString());
+            if (this.filterForJline(a)) {
+                return a;
+            }
+        }
+
+        super.getLog().warn("Unable to determine the required Jline dependency from the POM. Falling back to hard-coded defaults.");
+        super.getLog().warn("If you get an InvocationTargetException, then this probably means we guessed the wrong version for JLine");
+        super.getLog().warn(String.format("Guessed JLine: %s", defaultFallback.toString()));
+
+        return defaultFallback;
+    }
+
+    /**
+     * Helper function to filter a collection of {@link Artifact} for JLine.
+     * <p>
+     * Since different versions of the Scala Compiler were using different
+     * artifacts for JLine, things are a bit tricky here. Any {@link Artifact}
+     * to have an artifact id equal to {@code "jline"} and a group id equal to
+     * <em>either</em> {@code "jline"} or {@code "org.scala-lang"} will yield
+     * {@code true}. The latter group id corresponds to a fork of JLine that is
+     * not being used anymore.
+     *
+     * @param artifact the {@link Artifact} to check to see if it is viable
+     *        JLine candidate.
+     */
+    private boolean filterForJline(final Artifact artifact) {
+        final String artifactId = artifact.getArtifactId();
+        final String groupId = artifact.getGroupId();
+
+        if (artifactId.equals(ScalaConsoleMojo.JLINE) &&
+            (groupId.equals(ScalaConsoleMojo.JLINE) ||
+             groupId.equals(ScalaConsoleMojo.JLINE))) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Hard coded fallback values for JLine. This used to be the only way we
+     * resolve JLine, but required manually upkeep to avoid binary
+     * comparability errors.
+     * <p>
+     * You should favor {@link #resolveJLine} only using this for a fallback
+     * default.
+     * <p>
+     * Current mapping is as follows.
+     * <ul>
+     *   <li>Scala 2.12.0-M4 and after, jline:jline:2.14.1:jar</li>
+     *   <li>After Scala 2.11.0 through Scala 2.12.0-M3, jline:jline:2.12:jar</li>
+     *   <li>After Scala 2.9.0 before Scala 2.11.0, org.scala-lang:jline:SCALA-VERSION:jar</li>
+     *   <li>Before Scala 2.9.0, jline:jline:0.9.94:jar</li>
+     * </ul>
+     *
+     * @param scalaVersion the version of the Scala Compiler/Library we are
+     *        using for this execution.
+     *
+     * @return a fallback {@link Artifact} which we <em>hope</em> will work with
+     *         the Scala REPL.
+     */
+    private Artifact fallbackJLine(final VersionNumber scalaVersion) {
+        // https://github.com/scala/scala/blob/365ac035a863a666f86151371db77c6d401e88a2/versions.properties#L29
+        final VersionNumber scala2_12_0M4 = new VersionNumber("2.12.0-M4");
+        final VersionNumber scala2_11_0 = new VersionNumber("2.11.0");
+        final VersionNumber scala2_9_0 = new VersionNumber("2.9.0");
+
+        if (scala2_12_0M4.compareTo(scalaVersion) <= 0) {
+            return super.factory.createArtifact(ScalaConsoleMojo.JLINE, ScalaConsoleMojo.JLINE, "2.14.1", "", ScalaMojoSupport.JAR);
+        } else if (scala2_11_0.compareTo(scalaVersion) <= 0) {
+            return super.factory.createArtifact(ScalaConsoleMojo.JLINE, ScalaConsoleMojo.JLINE, "2.12", "", ScalaMojoSupport.JAR);
+        } else if (scala2_9_0.compareTo(scalaVersion) <= 0) {
+            return super.factory.createArtifact(ScalaConsoleMojo.SCALA_ORG_GROUP, ScalaConsoleMojo.JLINE, scalaVersion.toString(), "", ScalaMojoSupport.JAR);
+        } else {
+            return super.factory.createArtifact(ScalaConsoleMojo.JLINE, ScalaConsoleMojo.JLINE, "0.9.94", "", ScalaMojoSupport.JAR);
+        }
     }
 }

--- a/src/main/java/scala_maven/ScalaMojoSupport.java
+++ b/src/main/java/scala_maven/ScalaMojoSupport.java
@@ -8,6 +8,7 @@ import org.apache.maven.artifact.repository.ArtifactRepository;
 import org.apache.maven.artifact.resolver.ArtifactCollector;
 import org.apache.maven.artifact.resolver.ArtifactNotFoundException;
 import org.apache.maven.artifact.resolver.ArtifactResolutionException;
+import org.apache.maven.artifact.resolver.ArtifactResolutionRequest;
 import org.apache.maven.artifact.resolver.ArtifactResolver;
 import org.apache.maven.artifact.resolver.filter.AndArtifactFilter;
 import org.apache.maven.artifact.resolver.filter.ArtifactFilter;
@@ -51,6 +52,18 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
 
     public static final String SCALA_LIBRARY_ARTIFACTID= "scala-library";
     public static final String SCALA_COMPILER_ARTIFACTID= "scala-compiler";
+
+    /**
+     * Constant {@link String} for "pom". Used to specify the Maven POM artifact
+     * type.
+     */
+    protected static final String POM = "pom";
+
+    /**
+     * Constant {@link String} for "jar". Used to specify the Maven JAR artifact
+     * type.
+     */
+    protected static final String JAR = "jar";
 
     /**
      * @parameter property="project"
@@ -373,47 +386,136 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
     private VersionNumber _scalaVersionN;
 
     /**
-     * This method resolves the dependency artifacts from the project.
+     * Constructs an {@link Artifact} for Scala Compiler.
      *
-     * @param theProject The POM.
-     * @return resolved set of dependency artifacts.
+     * @param scalaVersion the version of the Scala Compiler/Library we are
+     *        using for this execution.
      *
-     * @throws ArtifactResolutionException
-     * @throws ArtifactNotFoundException
-     * @throws InvalidDependencyVersionException
+     * @return a {@link Artifact} for the Scala Compiler.
      */
-    protected Set<Artifact> resolveDependencyArtifacts(MavenProject theProject) throws Exception {
-        AndArtifactFilter filter = new AndArtifactFilter();
-        filter.add(new ScopeArtifactFilter(Artifact.SCOPE_TEST));
-        filter.add(new ArtifactFilter(){
-            public boolean include(Artifact artifact) {
-                return !artifact.isOptional();
-            }
-        });
-        //TODO follow the dependenciesManagement and override rules
-        Set<Artifact> artifacts = theProject.createArtifacts(artifactFactory, Artifact.SCOPE_RUNTIME, filter);
-        for (Artifact artifact : artifacts) {
-            resolver.resolve(artifact, remoteRepos, localRepo);
-        }
-        return artifacts;
+    protected final Artifact scalaCompilerArtifact(final String scalaVersion) {
+        return this.factory.createArtifact(this.getScalaOrganization(),
+                                           ScalaMojoSupport.SCALA_COMPILER_ARTIFACTID,
+                                           scalaVersion,
+                                           "",
+                                           ScalaMojoSupport.POM);
     }
 
     /**
      * This method resolves all transitive dependencies of an artifact.
      *
-     * @param artifact the artifact used to retrieve dependencies
+     * @param artifact the {@link Artifact} used to retrieve dependencies.
      *
-     * @return resolved set of dependencies
+     * @return resolved {@link Set} of dependencies.
      *
-     * @throws ArtifactResolutionException
-     * @throws ArtifactNotFoundException
-     * @throws ProjectBuildingException
-     * @throws InvalidDependencyVersionException
+     * @throws {@link Exception} when various artifact resolution mechanisms fail.
      */
-    protected Set<Artifact> resolveArtifactDependencies(Artifact artifact) throws Exception {
-        Artifact pomArtifact = factory.createArtifact(artifact.getGroupId(), artifact.getArtifactId(), artifact.getVersion(), "", "pom");
-        MavenProject pomProject = mavenProjectBuilder.buildFromRepository(pomArtifact, remoteRepos, localRepo);
-        return resolveDependencyArtifacts(pomProject);
+    protected final Set<Artifact> resolveArtifactDependencies(final Artifact artifact) throws Exception {
+        final AndArtifactFilter filter = new AndArtifactFilter();
+        filter.add(new ScopeArtifactFilter(Artifact.SCOPE_TEST));
+        filter.add(new ArtifactFilter(){
+                public boolean include(Artifact artifact) {
+                    return !artifact.isOptional();
+                }
+            });
+
+        // Use the collection filter as the resolution filter.
+        return resolveDependencyArtifacts(artifact,
+                                          filter,
+                                          filter);
+    }
+
+    /**
+     * This method resolves all transitive dependencies of an artifact.
+     *
+     * @param artifact the {@link Artifact} used to retrieve dependencies.
+     * @param collectionFilter an {@link ArtifactFilter} used to determine which
+     *        members of the dependency graph should be included in resolution.
+     * @param collectionFilter an {@link ArtifactFilter} used to determine which
+     *        members dependency graph should be downloaded.
+     *
+     * @return resolved {@link Set} of dependencies.
+     *
+     * @throws {@link Exception} when various artifact resolution mechanisms fail.
+     */
+    protected final Set<Artifact> resolveDependencyArtifacts(final Artifact artifact,
+                                                             final ArtifactFilter collectionFilter,
+                                                             final ArtifactFilter resolutionFilter) throws Exception {
+        return this.resolveDependencyArtifacts(artifact,
+                                               collectionFilter,
+                                               resolutionFilter,
+                                               this.remoteRepos,
+                                               this.localRepo);
+    }
+
+    /**
+     * This method resolves all transitive dependencies of an artifact.
+     *
+     * @param artifact the {@link Artifact} used to retrieve dependencies.
+     * @param collectionFilter an {@link ArtifactFilter} used to determine which
+     *        members of the dependency graph should be included in resolution.
+     * @param collectionFilter an {@link ArtifactFilter} used to determine which
+     *        members dependency graph should be downloaded.
+     * @param remoteRepositories a {@link List} of remote {@link
+     *        ArtifactRespository} values to used for dependency resolution of
+     *        the provided {@link Artifact}.
+     * @param localRepository the local {@link ArtifactRepository} to use for
+     *        dependency resolution of the given {@link Artifact}.
+     *
+     * @return resolved {@link Set} of dependencies.
+     *
+     * @throws {@link Exception} when various artifact resolution mechanisms fail.
+     */
+    protected final Set<Artifact> resolveDependencyArtifacts(final Artifact artifact,
+                                                             final ArtifactFilter collectionFilter,
+                                                             final ArtifactFilter resolutionFilter,
+                                                             final List<ArtifactRepository> remoteRepositories,
+                                                             final ArtifactRepository localRepository) throws Exception {
+        final ArtifactResolutionRequest arr =
+            this.createArtifactResolutionRequest(artifact,
+                                                 collectionFilter,
+                                                 resolutionFilter,
+                                                 remoteRepositories,
+                                                 localRepository);
+
+        //TODO follow the dependenciesManagement and override rules
+        return this.resolver.resolve(arr).getArtifacts();
+    }
+
+    /**
+     * Create a {@link ArtifactResolutionRequest}.
+     *
+     * @param artifact the {@link Artifact} used to retrieve dependencies.
+     * @param collectionFilter an {@link ArtifactFilter} used to determine which
+     *        members of the dependency graph should be included in resolution.
+     * @param collectionFilter an {@link ArtifactFilter} used to determine which
+     *        members dependency graph should be downloaded.
+     * @param remoteRepositories a {@link List} of remote {@link
+     *        ArtifactRespository} values to used for dependency resolution of
+     *        the provided {@link Artifact}.
+     * @param localRepository the local {@link ArtifactRepository} to use for
+     *        dependency resolution of the given {@link Artifact}.
+     *
+     * @return an {@link ArtifactResolutionRequest}, typically used for
+     *         dependency resolution requests against an {@link
+     *         ArtifactResolver}.
+     */
+    private ArtifactResolutionRequest createArtifactResolutionRequest(final Artifact artifact,
+                                                                      final ArtifactFilter collectionFilter,
+                                                                      final ArtifactFilter resolutionFilter,
+                                                                      final List<ArtifactRepository> remoteRepositories,
+                                                                      final ArtifactRepository localRepository) {
+        final ArtifactResolutionRequest arr = new ArtifactResolutionRequest();
+
+        arr.setArtifact(artifact);
+        arr.setCollectionFilter(collectionFilter);
+        arr.setResolutionFilter(resolutionFilter);
+        arr.setResolveRoot(false);
+        arr.setResolveTransitively(true);
+        arr.setRemoteRepositories(remoteRepositories);
+        arr.setLocalRepository(localRepository);
+
+        return arr;
     }
 
     public void addToClasspath(String groupId, String artifactId, String version, Set<String> classpath) throws Exception {
@@ -422,7 +524,7 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
 
 
     public void addToClasspath(String groupId, String artifactId, String version, Set<String> classpath, boolean addDependencies) throws Exception {
-        addToClasspath(factory.createArtifact(groupId, artifactId, version, Artifact.SCOPE_RUNTIME, "jar"), classpath, addDependencies);
+        addToClasspath(factory.createArtifact(groupId, artifactId, version, Artifact.SCOPE_RUNTIME, ScalaMojoSupport.JAR), classpath, addDependencies);
     }
 
     /**
@@ -435,7 +537,7 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
       d.setGroupId(groupId);
       d.setArtifactId(artifactId);
       d.setVersion(version);
-      d.setType("jar");
+      d.setType(ScalaMojoSupport.JAR);
       d.setClassifier(classifier);
       d.setScope(Artifact.SCOPE_RUNTIME);
       addToClasspath(factory.createDependencyArtifact(d), classpath, addDependencies);
@@ -502,7 +604,7 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
                 detectedScalaVersion = findScalaVersionFromDependencies();
             }
             if (StringUtils.isEmpty(detectedScalaVersion)) {
-                if (!"pom".equals( project.getPackaging().toLowerCase() )) {
+                if (!ScalaMojoSupport.POM.equals( project.getPackaging().toLowerCase() )) {
                     getLog().warn("you don't define "+ getScalaOrganization() + ":" + SCALA_LIBRARY_ARTIFACTID + " as a dependency of the project");
                 }
                 detectedScalaVersion = "0.0.0";
@@ -627,7 +729,28 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
 
 
     protected JavaMainCaller getScalaCommand() throws Exception {
-        JavaMainCaller cmd = getEmptyScalaCommand(scalaClassName);
+        return this.getScalaCommand(this.fork,
+                                    this.scalaClassName);
+    }
+
+    /**
+     * Get a {@link JavaMainCaller} used invoke a Java process. Typically this
+     * will be one of the Scala utilities (Compiler, ScalaDoc, REPL, etc.).
+     * <p>
+     * This method does some setup on the {@link JavaMainCaller} which is not
+     * done by merely invoking {@code new} on one of the
+     * implementations. Specifically, it adds any Scala compiler plugin options,
+     * JVM options, and Scalac options defined on the plugin.
+     *
+     * @param forkOverride override the setting for {@link #fork}. Currently
+     *        this should only be set if you are invoking the REPL.
+     * @param mainClass the JVM main class to invoke.
+     *
+     * @return a {@link JavaMainCaller} to use to invoke the given command.
+     */
+    protected final JavaMainCaller getScalaCommand(final boolean forkOverride,
+                                                   final String mainClass) throws Exception {
+        JavaMainCaller cmd = getEmptyScalaCommand(mainClass, forkOverride);
         cmd.addArgs(args);
         if (StringUtils.isNotEmpty(addScalacArgs)) {
           cmd.addArgs(StringUtils.split(addScalacArgs, "|"));
@@ -637,33 +760,63 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
         return cmd;
     }
 
-    protected JavaMainCaller getEmptyScalaCommand(String mainClass) throws Exception {
+    /**
+     * Get a {@link JavaMainCaller} used invoke a Java process. Typically this
+     * will be one of the Scala utilities (Compiler, ScalaDoc, REPL, etc.).
+     *
+     * @param mainClass the JVM main class to invoke.
+     *
+     * @return a {@link JavaMainCaller} to use to invoke the given command.
+     */
+    protected final JavaMainCaller getEmptyScalaCommand(final String mainClass) throws Exception {
+        return this.getEmptyScalaCommand(mainClass, this.fork);
+    }
 
-      //TODO - Fork or not depending on configuration?
-      JavaMainCaller cmd;
-      String toolcp = getToolClasspath();
-      if(fork) {
-        // HACK (better may need refactor)
-        boolean bootcp = true;
-        if (args != null) {
-          for(String arg : args) {
-            bootcp = bootcp && !"-nobootcp".equals(arg);
-          }
+    /**
+     * Get a {@link JavaMainCaller} used invoke a Java process. Typically this
+     * will be one of the Scala utilities (Compiler, ScalaDoc, REPL, etc.).
+     *
+     * @param mainClass the JVM main class to invoke.
+     * @param forkOverride override the setting for {@link #fork}. Currently
+     *        this should only be set if you are invoking the REPL.
+     *
+     * @return a {@link JavaMainCaller} to use to invoke the given command.
+     */
+    protected JavaMainCaller getEmptyScalaCommand(final String mainClass,
+                                                  final boolean forkOverride) throws Exception {
+
+        // If we are deviating from the plugin settings, let the user know
+        // what's going on.
+        if (forkOverride != this.fork) {
+            super.getLog().info("Fork behavior overridden");
+            super.getLog().info(String.format("Fork for this execution is %s.", String.valueOf(forkOverride)));
         }
-        String cp = bootcp ? "" : toolcp;
-        bootcp = bootcp && !(StringUtils.isNotEmpty(addScalacArgs) && addScalacArgs.contains("-nobootcp"));
-        // scalac with args in files
-        // * works only since 2.8.0
-        // * is buggy (don't manage space in path on windows)
-        getLog().debug("use java command with args in file forced : " + forceUseArgFile);
-        cmd = new JavaMainCallerByFork(this, mainClass, cp, null, null, forceUseArgFile, toolchainManager.getToolchainFromBuildContext("jdk", session));
-        if (bootcp) {
-          cmd.addJvmArgs("-Xbootclasspath/a:" + toolcp);
+
+        //TODO - Fork or not depending on configuration?
+        JavaMainCaller cmd;
+        String toolcp = getToolClasspath();
+        if(forkOverride) {
+            // HACK (better may need refactor)
+            boolean bootcp = true;
+            if (args != null) {
+                for(String arg : args) {
+                    bootcp = bootcp && !"-nobootcp".equals(arg);
+                }
+            }
+            String cp = bootcp ? "" : toolcp;
+            bootcp = bootcp && !(StringUtils.isNotEmpty(addScalacArgs) && addScalacArgs.contains("-nobootcp"));
+            // scalac with args in files
+            // * works only since 2.8.0
+            // * is buggy (don't manage space in path on windows)
+            getLog().debug("use java command with args in file forced : " + forceUseArgFile);
+            cmd = new JavaMainCallerByFork(this, mainClass, cp, null, null, forceUseArgFile, toolchainManager.getToolchainFromBuildContext("jdk", session));
+            if (bootcp) {
+                cmd.addJvmArgs("-Xbootclasspath/a:" + toolcp);
+            }
+        } else  {
+            cmd = new JavaMainCallerInProcess(this, mainClass, toolcp, null, null);
         }
-      } else  {
-        cmd = new JavaMainCallerInProcess(this, mainClass, toolcp, null, null);
-      }
-      return cmd;
+        return cmd;
     }
 
     private String getToolClasspath() throws Exception {
@@ -750,22 +903,21 @@ public abstract class ScalaMojoSupport extends AbstractMojo {
     }
 
     protected File getArtifactJar(String groupId, String artifactId, String version) throws Exception {
-        Artifact artifact = factory.createArtifact(groupId, artifactId, version, Artifact.SCOPE_RUNTIME, "jar");
+        Artifact artifact = factory.createArtifact(groupId, artifactId, version, Artifact.SCOPE_RUNTIME, ScalaMojoSupport.JAR);
         resolver.resolve(artifact, remoteRepos, localRepo);
         return artifact.getFile();
     }
 
     protected File getArtifactJar(String groupId, String artifactId, String version, String classifier) throws Exception {
-        Artifact artifact = factory.createArtifactWithClassifier(groupId, artifactId, version, "jar", classifier);
+        Artifact artifact = factory.createArtifactWithClassifier(groupId, artifactId, version, ScalaMojoSupport.JAR, classifier);
         resolver.resolve(artifact, remoteRepos, localRepo);
         return artifact.getFile();
     }
 
     protected Set<Artifact> getAllDependencies(String groupId, String artifactId, String version) throws Exception {
         Set<Artifact> result = new HashSet<Artifact>();
-        Artifact pom = factory.createArtifact(groupId, artifactId, version, "", "pom");
-        MavenProject p = mavenProjectBuilder.buildFromRepository(pom, remoteRepos, localRepo);
-        Set<Artifact> d = resolveDependencyArtifacts(p);
+        Artifact pom = factory.createArtifact(groupId, artifactId, version, "", ScalaMojoSupport.POM);
+        Set<Artifact> d = resolveArtifactDependencies(pom);
         result.addAll(d);
         for (Artifact dependency : d) {
             Set<Artifact> transitive = getAllDependencies(dependency.getGroupId(), dependency.getArtifactId(), dependency.getVersion());

